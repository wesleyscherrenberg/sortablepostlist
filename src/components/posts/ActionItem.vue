<script setup lang="ts">
import {computed} from "vue";
import type {ActionItemProp, ActionsType} from "@/types/Types";

const emit = defineEmits(['onTimeTravelChange'])

const props = defineProps<ActionItemProp>()

const onTimeTravelHandler = (action: ActionsType) => {
  emit('onTimeTravelChange', action)
}

const itemText = computed(() => {
  return `Moved post ${props.action.postId} from index ${props.action.fromIndex} to index ${props.action.toIndex}`
})
</script>

<template>
  <div class="p-4 flex items-center border-b border-theme-light last:border-none transform transition-all duration-500">
    <p class="flex-1" data-test="itemText">{{itemText}}</p>
    <button @click="onTimeTravelHandler(action)"
            aria-label="onTimeTravelBtn"
            role="button"
            data-test="onTimeTravelBtn"
            class="p-4 bg-theme-secondary py-3 px-4 flex-none rounded ">
      {{btnText}}
    </button>
  </div>
</template>
