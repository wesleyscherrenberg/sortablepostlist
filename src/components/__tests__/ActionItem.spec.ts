import {shallowMount, VueWrapper} from '@vue/test-utils';
import {beforeEach, describe, expect, it} from "vitest";
import ActionItem from "@/components/posts/ActionItem.vue";
import type {ActionsType} from "@/types/Types";

const factory = () => {
    return shallowMount(ActionItem, {
        props:  {
            action: {
                postId: 1,
                fromIndex: 0,
                toIndex: 1
            } as ActionsType,
            btnText: 'Click',
        }
    });
}

describe('ActionItem.vue', () => {

    let wrapper: VueWrapper;

    beforeEach(() => {
        wrapper = factory();
    });

    it('matches the snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot();
    });

    it('renders the component', () => {
        expect(wrapper.exists()).toBe(true);
    });

    it('renders the text correctly', () => {

        expect(wrapper.find('[data-test="itemText"]').text()).toBe('Moved post 1 from index 0 to index 1')
    })

    it('emits an event when the button is clicked', async () => {

        await wrapper.find('[data-test="onTimeTravelBtn"]').trigger('click');

        expect(wrapper.emitted()).toHaveProperty('onTimeTravelChange');
        expect(wrapper.emitted().onTimeTravelChange[0]).toEqual([{
            postId: 1,
            fromIndex: 0,
            toIndex: 1
        }]);
    });

    it('renders btnText properly', () => {
        expect(wrapper.find('[data-test="onTimeTravelBtn"]').text()).toBe('Click');
    });
});
