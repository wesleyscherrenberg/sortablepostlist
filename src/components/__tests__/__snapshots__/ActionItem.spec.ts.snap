// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ActionItem.vue > matches the snapshot 1`] = `
"<div class="p-4 flex items-center border-b border-theme-light last:border-none transform transition-all duration-500">
  <p class="flex-1" data-test="itemText">Moved post 1 from index 0 to index 1</p><button data-test="onTimeTravelBtn" class="p-4 bg-theme-secondary py-3 px-4 flex-none rounded">Click</button>
</div>"
`;
